README.md
# pidgin2adium [![Build Status][travis]](http://travis-ci.org/gabebw/pidgin2adium) [![Code Climate][codeclimate]](https://codeclimate.com/github/gabebw/pidgin2adium)

[travis]: https://travis-ci.org/gabebw/pidgin2adium.svg?branch=master
[codeclimate]: https://codeclimate.com/github/gabebw/pidgin2adium.svg

Convert [Pidgin](http://pidgin.im/) (formerly gaim) logs to the
[Adium](http://adium.im/) format. This is a command-line wrapper around the
[Pipio] log-parsing library.

[Pipio]: https://github.com/gabebw/pipio

## Install

    gem install pidgin2adium

## Quick Start

Let's say you have some logs in `~/pidgin-logs`, and your aliases are "Gabe
B-W", "Gabe", and "Gabe Berke-Williams". Then you should run this:

    pidgin2adium --in ~/pidgin-logs --aliases "Gabe B-W,Gabe,Gabe Berke-Williams"

By default, `pidgin2adium` outputs logs to the directory that Adium looks for
logs in. If you want to customize the output directory, use the `--out` option:

    pidgin2adium --in ~/pidgin-logs --aliases "Gabe B-W,Gabe" -o ./output-dir

## OK, what's with the aliases?

Pidgin2adium needs a comma-separated list of your aliases to work. Aliases make
it so that you show up in chats as (for example) `Me` instead of as
`best_screen_name_ever_018845`.

Pidgin then uses aliases in the log file instead of the actual screen name,
which makes it impossible to match "Me" to your actual screen name.  Therefore
Pidgin2adium needs to know which aliases belong to you so it can map them to the
correct screen name.

If Pidgin2adium encounters an alias that you did not list, it assumes that it
belongs to the person to whom you are chatting.

You do not need to provide your screenname in the alias list.

## Testing

To get a coverage report, run `rake` with the `COVERAGE` environment variable
set:

    COVERAGE=1 rake

## THANKS

With thanks to Li Ma, whose [blog post](http://li-ma.blogspot.com/2008/10/pidgin-log-file-to-adium-log-converter.html)
helped tremendously.

## Note on Patches/Pull Requests

* Fork the project.
* Make your feature addition or bug fix.
* Add tests for it. This is important so I don't break it in a
  future version unintentionally.
* Commit, do not mess with rakefile, version, or history.
  (if you want to have your own version, that is fine but bump version in a commit by itself I can ignore when I pull)
* Send me a pull request. Bonus points for topic branches.

## Copyright

Copyright (c) 2009-2014 Gabe Berke-Williams. See LICENSE for details.


NEWS.md
### HEAD (unreleased)

* Users can specify a custom output directory with `-o`/`--out`
* Extract `ArgumentParser` class
* Drop support for Ruby 1.9.x (#20)

### 4.0.0.beta2

* Extract `Pidgin2Adium::Cli`
* Fix bugs
* Print error and progress messages

### 4.0.0.beta1

* Massive refactoring
* Use Pipio gem to parse logs
* Chat times respect time zones instead of assuming inputs are UTC
* Test against Travis

### 3.3.0 / 2011-10-16
* Pidgin2Adium depended on itself for some reason. Now it doesn't.

### 3.2.3 / 2010-11-08
* Be more liberal in what Pidgin2Adium accepts when parsing the date and
  getting basic time info for the first line. Thanks to Matthew Jakubowski for
  helping me fix this bug.

### 3.2.2 / 2010-11-08
* Use DateTime#strftime to get dates in xmlschema format. DateTime#xmlschema
  doesn't exist in Ruby 1.8, and Ruby 1.9 has DateTime#iso8601, not
  DateTime#xmlschema. Just use strftime. Thanks to Matthew Jakubowski for
  pointing this bug out.

### 3.2.1 / 2010-11-08
* Use straight `DateTime.parse` when possible, and only fall back on hacky
  `Date._strptime` when we have to.

### 3.2.0 / 2010-10-12
* Last release broke 1.8 compatibility due to use of strptime. 1.8 and 1.9
  both work in 3.2.0.
* Moved Pidgin2Adium::VERSION to its own file

### 3.1.1 / 2010-08-13
* Moved BasicParser and its subclasses into parsers/ subdir.
  - You can now do `require 'pidgin2adium/parsers/all'`,
    though the old `require 'pidgin2adium/log_parser'` will still work
* Moved Message and its subclasses into messages/ subdir
  - You can now do `require 'pidgin2adium/messages/all'`,
    though the old `require 'pidgin2adium/message'` will still work

### 3.1.0 / 2010-08-13
* Compatible with Ruby 1.9!
  - removed dependency on `parsedate` library, which 1.9 doesn't have
* `log_parser.rb` has been split into separate files (1 per class, more or less)
  - `require pidgin2adium/log_parser` will still pull in all of the split-up
    classes
* `balance_tags_c` extension really does work now
* Cleans up more junk from Pidgin logfiles when parsing
* Bugfixes and more graceful handling of error states
* Fully tested (except bin/pidgin2adium, which remains tricky)

### 3.0.1 / 2010-08-07
Bugfix release.

* `balance_tags_c.c`: Use `rb_eval_string` instead of `rb_reg_regcomp` to avoid
  segfaults (commit #733ce88b0836256e14f0, fixes #27811)

Non-user-facing stuff:

* Switched to Jeweler, RSpec, and Bundler
* Rakefile now doesn't choke if Hanna gem isn't installed

### 3.0.0 / 2010-01-31
* `lbalance_tags.rb` is now a C extension (`Pidgin2Adium.balance_tags_c`)
  - the pure-ruby mixin `balance_tags` (without the trailing `_c`) is gone
* Better handling of command-line arguments
* Format time zones offsets correctly (e.g. "+0500", not "+-0500")
* Write Yahoo! and Jabber logs to correct directories (#27710)
* Better matching of regexes against time strings
* Better documentation

### 2.0.2 / 2009-12-18
* Much better documentation (more of it, and higher quality too!)
* Allow user-provided output dir at commandline
* Cleaner error messages
* require 'time' for `Time.zone_offset` (fixes bug)
* Gracefully handle lack of timezone info
* Gracefully handle nonexistent output dir
* Gracefully handle parsing errors
* Print error messages during *and* after batch converting so they're actually seen

### 2.0.1 / 2009-12-06
* Fix timestamps so they show up in Adium chat log viewer

### 2.0.0 / 2009-11-24
* Added documentation, available at http://pidgin2adium.rubyforge.org/rdoc/
* Added public interface for scripting purposes
* Removed -o and -l options. Now gem automatically outputs to Adium log dir with no intermediate folder.

### 1.0.0 / 2009-09-27
* Birthday!


ragel/thing.rb

# line 1 "thing.rl"

# line 31 "thing.rl"



# line 9 "thing.rb"
class << self
	attr_accessor :_thing_actions
	private :_thing_actions, :_thing_actions=
end
self._thing_actions = [
	0, 1, 0, 1, 1, 1, 2, 1, 
	3, 1, 4, 1, 5, 1, 6
]

class << self
	attr_accessor :_thing_key_offsets
	private :_thing_key_offsets, :_thing_key_offsets=
end
self._thing_key_offsets = [
	0, 0, 1, 2, 3, 4, 5, 6, 
	7, 8, 9, 10, 11, 12, 13, 14, 
	15, 16, 17, 18, 19, 20, 21, 22, 
	23, 24, 25, 26, 27, 28, 29, 30, 
	31, 32, 33, 34, 35, 36, 37, 38, 
	39, 40, 41, 42, 43, 44, 45, 46, 
	47, 48, 49, 50, 51, 52, 53, 54, 
	55, 56, 57, 58, 59, 60, 61, 62, 
	63, 64, 65, 66, 67, 68, 69, 70, 
	71, 72, 73, 74, 75, 76, 77, 78, 
	79, 80, 81, 82, 83, 84, 85, 86, 
	87, 88, 89, 90, 91, 92, 93, 94, 
	95, 96, 97, 106, 107, 108, 117, 125
]

class << self
	attr_accessor :_thing_trans_keys
	private :_thing_trans_keys, :_thing_trans_keys=
end
self._thing_trans_keys = [
	104, 101, 97, 100, 62, 60, 109, 101, 
	116, 97, 32, 104, 116, 116, 112, 45, 
	101, 113, 117, 105, 118, 61, 34, 99, 
	111, 110, 116, 101, 110, 116, 45, 116, 
	121, 112, 101, 34, 32, 99, 111, 110, 
	116, 101, 110, 116, 61, 34, 116, 101, 
	120, 116, 47, 104, 116, 109, 108, 59, 
	32, 99, 104, 97, 114, 115, 101, 116, 
	61, 85, 84, 70, 45, 56, 34, 62, 
	60, 116, 105, 116, 108, 101, 62, 67, 
	111, 110, 118, 101, 114, 115, 97, 116, 
	105, 111, 110, 32, 119, 105, 116, 104, 
	32, 32, 37, 43, 48, 57, 65, 90, 
	97, 122, 97, 116, 37, 43, 60, 48, 
	57, 65, 90, 97, 122, 37, 43, 48, 
	57, 65, 90, 97, 122, 37, 43, 48, 
	57, 65, 90, 97, 122, 0
]

class << self
	attr_accessor :_thing_single_lengths
	private :_thing_single_lengths, :_thing_single_lengths=
end
self._thing_single_lengths = [
	0, 1, 1, 1, 1, 1, 1, 1, 
	1, 1, 1, 1, 1, 1, 1, 1, 
	1, 1, 1, 1, 1, 1, 1, 1, 
	1, 1, 1, 1, 1, 1, 1, 1, 
	1, 1, 1, 1, 1, 1, 1, 1, 
	1, 1, 1, 1, 1, 1, 1, 1, 
	1, 1, 1, 1, 1, 1, 1, 1, 
	1, 1, 1, 1, 1, 1, 1, 1, 
	1, 1, 1, 1, 1, 1, 1, 1, 
	1, 1, 1, 1, 1, 1, 1, 1, 
	1, 1, 1, 1, 1, 1, 1, 1, 
	1, 1, 1, 1, 1, 1, 1, 1, 
	1, 1, 3, 1, 1, 3, 2, 2
]

class << self
	attr_accessor :_thing_range_lengths
	private :_thing_range_lengths, :_thing_range_lengths=
end
self._thing_range_lengths = [
	0, 0, 0, 0, 0, 0, 0, 0, 
	0, 0, 0, 0, 0, 0, 0, 0, 
	0, 0, 0, 0, 0, 0, 0, 0, 
	0, 0, 0, 0, 0, 0, 0, 0, 
	0, 0, 0, 0, 0, 0, 0, 0, 
	0, 0, 0, 0, 0, 0, 0, 0, 
	0, 0, 0, 0, 0, 0, 0, 0, 
	0, 0, 0, 0, 0, 0, 0, 0, 
	0, 0, 0, 0, 0, 0, 0, 0, 
	0, 0, 0, 0, 0, 0, 0, 0, 
	0, 0, 0, 0, 0, 0, 0, 0, 
	0, 0, 0, 0, 0, 0, 0, 0, 
	0, 0, 3, 0, 0, 3, 3, 3
]

class << self
	attr_accessor :_thing_index_offsets
	private :_thing_index_offsets, :_thing_index_offsets=
end
self._thing_index_offsets = [
	0, 0, 2, 4, 6, 8, 10, 12, 
	14, 16, 18, 20, 22, 24, 26, 28, 
	30, 32, 34, 36, 38, 40, 42, 44, 
	46, 48, 50, 52, 54, 56, 58, 60, 
	62, 64, 66, 68, 70, 72, 74, 76, 
	78, 80, 82, 84, 86, 88, 90, 92, 
	94, 96, 98, 100, 102, 104, 106, 108, 
	110, 112, 114, 116, 118, 120, 122, 124, 
	126, 128, 130, 132, 134, 136, 138, 140, 
	142, 144, 146, 148, 150, 152, 154, 156, 
	158, 160, 162, 164, 166, 168, 170, 172, 
	174, 176, 178, 180, 182, 184, 186, 188, 
	190, 192, 194, 201, 203, 205, 212, 218
]

class << self
	attr_accessor :_thing_indicies
	private :_thing_indicies, :_thing_indicies=
end
self._thing_indicies = [
	0, 1, 2, 1, 3, 1, 4, 1, 
	5, 1, 6, 1, 7, 1, 8, 1, 
	9, 1, 10, 1, 11, 1, 12, 1, 
	13, 1, 14, 1, 15, 1, 16, 1, 
	17, 1, 18, 1, 19, 1, 20, 1, 
	21, 1, 22, 1, 23, 1, 24, 1, 
	25, 1, 26, 1, 27, 1, 28, 1, 
	29, 1, 30, 1, 31, 1, 32, 1, 
	33, 1, 34, 1, 35, 1, 36, 1, 
	37, 1, 38, 1, 39, 1, 40, 1, 
	41, 1, 42, 1, 43, 1, 44, 1, 
	45, 1, 46, 1, 47, 1, 48, 1, 
	49, 1, 50, 1, 51, 1, 52, 1, 
	53, 1, 54, 1, 55, 1, 56, 1, 
	57, 1, 58, 1, 59, 1, 60, 1, 
	61, 1, 62, 1, 63, 1, 64, 1, 
	65, 1, 66, 1, 67, 1, 68, 1, 
	69, 1, 70, 1, 71, 1, 72, 1, 
	73, 1, 74, 1, 75, 1, 76, 1, 
	77, 1, 78, 1, 79, 1, 80, 1, 
	81, 1, 82, 1, 83, 1, 84, 1, 
	85, 1, 86, 1, 87, 1, 88, 1, 
	89, 1, 90, 1, 91, 1, 92, 1, 
	93, 1, 94, 1, 95, 1, 96, 1, 
	97, 1, 99, 100, 100, 100, 100, 100, 
	98, 101, 98, 102, 98, 103, 103, 104, 
	103, 103, 103, 1, 103, 103, 103, 103, 
	103, 105, 100, 100, 100, 100, 100, 106, 
	0
]

class << self
	attr_accessor :_thing_trans_targs
	private :_thing_trans_targs, :_thing_trans_targs=
end
self._thing_trans_targs = [
	2, 0, 3, 4, 5, 6, 7, 8, 
	9, 10, 11, 12, 13, 14, 15, 16, 
	17, 18, 19, 20, 21, 22, 23, 24, 
	25, 26, 27, 28, 29, 30, 31, 32, 
	33, 34, 35, 36, 37, 38, 39, 40, 
	41, 42, 43, 44, 45, 46, 47, 48, 
	49, 50, 51, 52, 53, 54, 55, 56, 
	57, 58, 59, 60, 61, 62, 63, 64, 
	65, 66, 67, 68, 69, 70, 71, 72, 
	73, 74, 75, 76, 77, 78, 79, 80, 
	81, 82, 83, 84, 85, 86, 87, 88, 
	89, 90, 91, 92, 93, 94, 95, 96, 
	97, 103, 101, 99, 98, 100, 101, 102, 
	1, 101, 101
]

class << self
	attr_accessor :_thing_trans_actions
	private :_thing_trans_actions, :_thing_trans_actions=
end
self._thing_trans_actions = [
	0, 0, 0, 0, 0, 0, 0, 0, 
	0, 0, 0, 0, 0, 0, 0, 0, 
	0, 0, 0, 0, 0, 0, 0, 0, 
	0, 0, 0, 0, 0, 0, 0, 0, 
	0, 0, 0, 0, 0, 0, 0, 0, 
	0, 0, 0, 0, 0, 0, 0, 0, 
	0, 0, 0, 0, 0, 0, 0, 0, 
	0, 0, 0, 0, 0, 0, 0, 0, 
	0, 0, 0, 0, 0, 0, 0, 0, 
	0, 0, 0, 0, 0, 0, 0, 0, 
	0, 0, 0, 0, 0, 0, 0, 0, 
	0, 0, 0, 0, 0, 0, 0, 0, 
	0, 5, 13, 0, 0, 0, 7, 0, 
	0, 11, 9
]

class << self
	attr_accessor :_thing_to_state_actions
	private :_thing_to_state_actions, :_thing_to_state_actions=
end
self._thing_to_state_actions = [
	0, 0, 0, 0, 0, 0, 0, 0, 
	0, 0, 0, 0, 0, 0, 0, 0, 
	0, 0, 0, 0, 0, 0, 0, 0, 
	0, 0, 0, 0, 0, 0, 0, 0, 
	0, 0, 0, 0, 0, 0, 0, 0, 
	0, 0, 0, 0, 0, 0, 0, 0, 
	0, 0, 0, 0, 0, 0, 0, 0, 
	0, 0, 0, 0, 0, 0, 0, 0, 
	0, 0, 0, 0, 0, 0, 0, 0, 
	0, 0, 0, 0, 0, 0, 0, 0, 
	0, 0, 0, 0, 0, 0, 0, 0, 
	0, 0, 0, 0, 0, 0, 0, 0, 
	0, 0, 0, 0, 0, 1, 0, 0
]

class << self
	attr_accessor :_thing_from_state_actions
	private :_thing_from_state_actions, :_thing_from_state_actions=
end
self._thing_from_state_actions = [
	0, 0, 0, 0, 0, 0, 0, 0, 
	0, 0, 0, 0, 0, 0, 0, 0, 
	0, 0, 0, 0, 0, 0, 0, 0, 
	0, 0, 0, 0, 0, 0, 0, 0, 
	0, 0, 0, 0, 0, 0, 0, 0, 
	0, 0, 0, 0, 0, 0, 0, 0, 
	0, 0, 0, 0, 0, 0, 0, 0, 
	0, 0, 0, 0, 0, 0, 0, 0, 
	0, 0, 0, 0, 0, 0, 0, 0, 
	0, 0, 0, 0, 0, 0, 0, 0, 
	0, 0, 0, 0, 0, 0, 0, 0, 
	0, 0, 0, 0, 0, 0, 0, 0, 
	0, 0, 0, 0, 0, 3, 0, 0
]

class << self
	attr_accessor :_thing_eof_trans
	private :_thing_eof_trans, :_thing_eof_trans=
end
self._thing_eof_trans = [
	0, 0, 0, 0, 0, 0, 0, 0, 
	0, 0, 0, 0, 0, 0, 0, 0, 
	0, 0, 0, 0, 0, 0, 0, 0, 
	0, 0, 0, 0, 0, 0, 0, 0, 
	0, 0, 0, 0, 0, 0, 0, 0, 
	0, 0, 0, 0, 0, 0, 0, 0, 
	0, 0, 0, 0, 0, 0, 0, 0, 
	0, 0, 0, 0, 0, 0, 0, 0, 
	0, 0, 0, 0, 0, 0, 0, 0, 
	0, 0, 0, 0, 0, 0, 0, 0, 
	0, 0, 0, 0, 0, 0, 0, 0, 
	0, 0, 0, 0, 0, 0, 0, 0, 
	0, 0, 99, 99, 99, 0, 106, 107
]

class << self
	attr_accessor :thing_start
end
self.thing_start = 101;
class << self
	attr_accessor :thing_first_final
end
self.thing_first_final = 101;
class << self
	attr_accessor :thing_error
end
self.thing_error = 0;

class << self
	attr_accessor :thing_en_main
end
self.thing_en_main = 101;


# line 34 "thing.rl"

def emit(token_name, data, target_array, ts, te)
  target_array << {name: token_name.to_sym, value: data[ts...te].pack("c*") }
end

def run_lexer(data)
  data = data.unpack("c*") if(data.is_a?(String))
  eof = data.length
  token_array = []

  
# line 292 "thing.rb"
begin
	p ||= 0
	pe ||= data.length
	cs = thing_start
	ts = nil
	te = nil
	act = 0
end

# line 45 "thing.rl"
  
# line 304 "thing.rb"
begin
	_klen, _trans, _keys, _acts, _nacts = nil
	_goto_level = 0
	_resume = 10
	_eof_trans = 15
	_again = 20
	_test_eof = 30
	_out = 40
	while true
	_trigger_goto = false
	if _goto_level <= 0
	if p == pe
		_goto_level = _test_eof
		next
	end
	if cs == 0
		_goto_level = _out
		next
	end
	end
	if _goto_level <= _resume
	_acts = _thing_from_state_actions[cs]
	_nacts = _thing_actions[_acts]
	_acts += 1
	while _nacts > 0
		_nacts -= 1
		_acts += 1
		case _thing_actions[_acts - 1]
			when 1 then
# line 1 "NONE"
		begin
ts = p
		end
# line 338 "thing.rb"
		end # from state action switch
	end
	if _trigger_goto
		next
	end
	_keys = _thing_key_offsets[cs]
	_trans = _thing_index_offsets[cs]
	_klen = _thing_single_lengths[cs]
	_break_match = false
	
	begin
	  if _klen > 0
	     _lower = _keys
	     _upper = _keys + _klen - 1

	     loop do
	        break if _upper < _lower
	        _mid = _lower + ( (_upper - _lower) >> 1 )

	        if data[p].ord < _thing_trans_keys[_mid]
	           _upper = _mid - 1
	        elsif data[p].ord > _thing_trans_keys[_mid]
	           _lower = _mid + 1
	        else
	           _trans += (_mid - _keys)
	           _break_match = true
	           break
	        end
	     end # loop
	     break if _break_match
	     _keys += _klen
	     _trans += _klen
	  end
	  _klen = _thing_range_lengths[cs]
	  if _klen > 0
	     _lower = _keys
	     _upper = _keys + (_klen << 1) - 2
	     loop do
	        break if _upper < _lower
	        _mid = _lower + (((_upper-_lower) >> 1) & ~1)
	        if data[p].ord < _thing_trans_keys[_mid]
	          _upper = _mid - 2
	        elsif data[p].ord > _thing_trans_keys[_mid+1]
	          _lower = _mid + 2
	        else
	          _trans += ((_mid - _keys) >> 1)
	          _break_match = true
	          break
	        end
	     end # loop
	     break if _break_match
	     _trans += _klen
	  end
	end while false
	_trans = _thing_indicies[_trans]
	end
	if _goto_level <= _eof_trans
	cs = _thing_trans_targs[_trans]
	if _thing_trans_actions[_trans] != 0
		_acts = _thing_trans_actions[_trans]
		_nacts = _thing_actions[_acts]
		_acts += 1
		while _nacts > 0
			_nacts -= 1
			_acts += 1
			case _thing_actions[_acts - 1]
when 2 then
# line 1 "NONE"
		begin
te = p+1
		end
when 3 then
# line 27 "thing.rl"
		begin
te = p+1
 begin 
      emit(:first_line, data, token_array, ts, te)
     end
		end
when 4 then
# line 19 "thing.rl"
		begin
te = p
p = p - 1; begin 
      emit(:pre, data, token_array, ts, te)
     end
		end
when 5 then
# line 23 "thing.rl"
		begin
te = p
p = p - 1; begin 
      emit(:username, data, token_array, ts, te)
     end
		end
when 6 then
# line 19 "thing.rl"
		begin
 begin p = ((te))-1; end
 begin 
      emit(:pre, data, token_array, ts, te)
     end
		end
# line 442 "thing.rb"
			end # action switch
		end
	end
	if _trigger_goto
		next
	end
	end
	if _goto_level <= _again
	_acts = _thing_to_state_actions[cs]
	_nacts = _thing_actions[_acts]
	_acts += 1
	while _nacts > 0
		_nacts -= 1
		_acts += 1
		case _thing_actions[_acts - 1]
when 0 then
# line 1 "NONE"
		begin
ts = nil;		end
# line 462 "thing.rb"
		end # to state action switch
	end
	if _trigger_goto
		next
	end
	if cs == 0
		_goto_level = _out
		next
	end
	p += 1
	if p != pe
		_goto_level = _resume
		next
	end
	end
	if _goto_level <= _test_eof
	if p == eof
	if _thing_eof_trans[cs] > 0
		_trans = _thing_eof_trans[cs] - 1;
		_goto_level = _eof_trans
		next;
	end
end
	end
	if _goto_level <= _out
		break
	end
	end
	end

# line 46 "thing.rl"

  puts token_array.inspect
end

string = '<head><meta http-equiv="content-type" content="text/html; charset=UTF-8"><title>Conversation with aolsystemmsg at 10/13/2007 12:47:52 PM on jiggerificbug (aim)</title></head><h3>Conversation with aolsystemmsg at 10/13/2007 12:47:52 PM on jiggerificbug (aim)</h3>'
run_lexer(string)
